jaxlib/rocblas.cc

```diff
@@ -0,0 +1,724 @@
+/* Copyright 2019 Google LLC
```

Comment:
My main concern with this file is the amount of code duplication with the cusolver implementation.

I'm wondering if we would do well to introduce a thin interface abstraction over either solver library, but keep the kernel code otherwise shared. This would also allow us to avoid a number of the changes at higher levels: e.g., we wouldn't need to clone the python bindings or change many of the translation rules.

The only significant difference I noticed is that rocsolver doesn't return `info` values in some cases. Presumably we could handle that by returning an `info` value of `0` ("ok")?

What do you think?

Reply:
something like https://github.com/tensorflow/tensorflow/blob/6859f52a3fba6714b5360262f190c9649613ac5c/tensorflow/stream_executor/blas.h ?

I think it would be very nice, however there are quite a few other places where they differ: e.g. in cublas/rocblas there are two separate versions of Getrf. Also I added some more batched versions from rocsolver directly where the cuda verision still uses loops (although it might be possible to add them, I have not checked...). Also in gesvd the buffers work a bit differently.

But I think it should at least be possible to share the python part, adding a few if statements here and there.

Reply:
On reflection, I think the right thing to do here would be to move these kernels into XLA itself, so, e.g., TF can share them. Let's therefore not spend too much time on adding better abstractions around them here.